# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos:  Pedro Luis Soto Santos <br>
Url: https://github.com/plss12/03MAIR---Algoritmos-de-Optimizacion/blob/main/Trabajo/Seminario_Algoritmos.ipynb<br>

Problema:
>3. Combinar cifras y operaciones

Descripción del problema:

El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.

Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/).

Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad dada. Un ejemplo sería para obtener el 4:
4+2-6/3*1 = 4

Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las siguientes cuestiones:
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?
- ¿Es posible encontrar t odos los valores enteros posibles entre dicho mínimo y máximo ?

                                        

In [39]:
# Para las preguntas que se plantean sobre el valor máximo y mínimo calculable con estas condiciones 
# del problema y, dados este máximo y este mínimo, saber si podemos calcular todos los valores enteros 
# entre ellos, he diseñado el siguiente código:

from itertools import permutations

def numeros_posibles():

    numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    operadores = ['+', '-', '*', '/']
    numeros_expresiones = {}

    # Generamos todas las permutaciones de 5 números y 4 operadores
    permuts_numeros = permutations(numeros, 5)
    for numeros_permutados in permuts_numeros:
        permuts_operadores = permutations(operadores)
        for operadores_permutados in permuts_operadores:
            expresion = []
            
            # Unimos los números y los operadores en una lista y formateamos a String
            for i in range(4):
                expresion.append(str(numeros_permutados[i]))
                expresion.append(operadores_permutados[i])
            expresion.append(str(numeros_permutados[4]))
            expresion_str = ''.join(expresion)

            # Evaluamos la expresión y guardamos el resultado si es un entero y no se ha obtenido antes
            resultado = eval(expresion_str)
            if resultado.is_integer() and resultado not in numeros_expresiones.keys():
                numeros_expresiones[resultado] = expresion_str
                
    return numeros_expresiones

# Cálculamos la lista completa de números enteros que se pueden obtener con las restricciones del problema
numeros = numeros_posibles()

print(f"Números enteros calculados: {len(numeros)}" )

# Obetenemos el máximo y el mínimo de la lista, como se nos pide en el problema
num_max = max(numeros)
num_min = min(numeros)

print(f"Número entero Máximo: {num_max}")
print(f"Número entero Mínimo: {num_min}")

# A partir del máximo, el mínimo y la cantidad de números obtenidos, podemos comprobar si se pueden obtener
# todos los números enteros entre el mínimo y el máximo, como se nos pide en el problema
if num_max - num_min + 1 == len(numeros):
    print(f"Se pueden obtener todos los números enteros entre {num_min} y {num_max}")
else:
    print(f"No se pueden obtener todos los números enteros entre {num_min} y {num_max}")

Números enteros calculados: 147
Número entero Máximo: 77.0
Número entero Mínimo: -69.0
Se pueden obtener todos los números enteros entre -69.0 y 77.0


(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>



¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.




RESPUESTA:

El problema dice que se deben combinar los números y los operadores sin repetir ninguno de ellos. Esta restricción provoca que al haber tan solo 4 operadores tengamos una operación limitada a 5 números. Por lo que, si no tenemos en cuenta la restricción de no repetición de números ni operadores, tenemos infintas posibilidades a la hora de formar operaciones.

Sin embargo, teniendo en cuenta las restricciones podemos calcular el número total de posibles operaciones, sabiendo que habrá 5 números y 4 operadores. En otras palabras, sería el resultado de calcular una variación sin repetición cogiendo 5 valores de los 9 posibles números multiplicado por una permutación sin repetición para el caso de los 4 operadores:

$$V(9,5) = \frac{9!}{(9-5)!} = 15120$$

$$P(4) = 4! = 24$$

$$Total = 15120 \cdot 24 = 362880$$

Este resultado ha sido calculado a priori y además ha sido calculado con el código de python de abajo, que es una modificación del código anterior, en el que solo tenemos en cuenta las expresiones generadas.

In [40]:
def operaciones_posibles():

    numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    operadores = ['+', '-', '*', '/']
    expresiones = []

    # Generamos todas las permutaciones de 5 números y 4 operadores
    permuts_numeros = permutations(numeros, 5)
    for numeros_permutados in permuts_numeros:
        permuts_operadores = permutations(operadores)
        for operadores_permutados in permuts_operadores:
            expresion = []
            
            # Unimos los números y los operadores en una lista y formateamos a String
            for i in range(4):
                expresion.append(str(numeros_permutados[i]))
                expresion.append(operadores_permutados[i])
            expresion.append(str(numeros_permutados[4]))
            expresion_str = ''.join(expresion)
            expresiones.append(expresion_str)
                
    return expresiones

operaciones = operaciones_posibles()
print(f"Número de posibles operaciones con restricciones: {len(operaciones)}")

Número de posibles operaciones con restricciones: 362880


Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


RESPUESTA:

Se ha decidido utilizar una lista para almacenar los números y los operadores para poder ser usados por la función permutations, de la biblioteca itertools. 

Para el resultado simplemente se devuelve la operación, que da como resultado el valor objetivo, como un string y en el caso de que no exista dicha operación se devuelve un string vacío.

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

RESPUESTA:

Este problema no es un claro problema de optimización, o al menos yo no lo he planteado como tal, por lo que no creo que haya una función objetivo clara, ya que lo que se está buscando en este problema es una operación que de como resultado un valor dado como objetivo. 

Sin embargo, se podría decir que una función objetivo sería encontrar la operación que da como resultado este valor objetivo en el menor número de iteraciones posibles, siendo en este caso un problema de minimización.

Diseña un algoritmo para resolver el problema por fuerza bruta

RESPUESTA:

En este caso la función que resuelve el problema por fuerza bruta es el código de la función mostrada al principio, con la modificación de que se le pasa el valor objetivo que se está buscando, y si se encuentra una operación que da como resultado este valor directamente se devuelve. Además, he añadido un contador de iteraciones, para más adelante hacer comparaciones con el algoritmo más eficiente.

In [41]:
def obtener_expresion(numero):

    numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    operadores = ['+', '-', '*', '/']

    iteraciones = 0

    # Generamos todas las permutaciones de 5 números y 4 operadores
    permuts_numeros = permutations(numeros, 5)
    for numeros_permutados in permuts_numeros:
        permuts_operadores = permutations(operadores)
        for operadores_permutados in permuts_operadores:
            expresion = []
            
            # Unimos los números y los operadores en una lista y formateamos a String
            for i in range(4):
                expresion.append(str(numeros_permutados[i]))
                expresion.append(operadores_permutados[i])
            expresion.append(str(numeros_permutados[4]))
            expresion_str = ''.join(expresion)
            
            iteraciones += 1

            # Evaluamos la expresión y guardamos el resultado si es un entero y no se ha obtenido antes
            resultado = eval(expresion_str)
            if resultado == numero:
                return expresion_str, iteraciones
            
    return "", iteraciones

#Función para no repetir código a la hora de imprimir el resultado
def imprimir_resultado(operacion, iteraciones, numero_objetivo):
    print(f"Realizadas {iteraciones} iteraciones para obtener el número {numero_objetivo}")
    if operacion != "":
        print(f"Expresión para obtener el número {numero_objetivo}: {operacion}\n")
    else:
        print(f"Dadas las restricciones del problema, no existe una expresión para obtener el número {numero_objetivo}\n")

# Ponemos un ejemplo en el que sí se puede obtener el número objetivo y otro en el que no
numero_objetivo = 12
expresion, iteraciones = obtener_expresion(numero_objetivo)
imprimir_resultado(expresion, iteraciones, numero_objetivo)

numero_objetivo = 200
expresion, iteraciones = obtener_expresion(numero_objetivo)
imprimir_resultado(expresion, iteraciones, numero_objetivo)

Realizadas 3483 iteraciones para obtener el número 12
Expresión para obtener el número 12: 1+2*7-9/3

Realizadas 362880 iteraciones para obtener el número 200
Dadas las restricciones del problema, no existe una expresión para obtener el número 200



Calcula la complejidad del algoritmo por fuerza bruta

RESPUESTA:

Este algoritmo por fuerza bruta evalua todas las combinaciones de operadores y números hasta encontrar la operación que como resultado da el número objetivo. En otras palabras, el algoritmo cuenta con dos for anidados, uno recorre todas las variaciones de 5 números y el otro recorre todas las permutaciones de los operadores. Además, hay un pequeño for final que monta la expresión, pero este al ser un for constante su complejidad también es constante.

Usando las formulas de permutación y variación vistas anteriormente:

$$primerfor = O(\frac{n!}{(n-5)!})$$

$$segundofor = O(m!)$$

$$tercerfor = O(1)$$

$$complejidad = O(\frac{n!}{(n-5)!}) \cdot O(m!)$$


(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

RESPUESTA:

Para mejorar la complejidad del algoritmo por fuerza bruta he optado por un algoritmo de backtracking con llamadas recursivas, que irá construyendo paso a paso las operaciones, evitando tener que crear y recorrer todas las posibles operaciones con las permutaciones de los números y los operadores. Además, he decidido ir evaluando las expresión conforme se iba montando para descartar las ramas en las que se obtenía como resultado parcial un número decimal, ya que es muy poco probable que a partir de un número decimal se obtenga como resultado final un número entero.

In [42]:
def obtener_expresion_backtrack(objetivo):
    numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    operadores = ['+', '-', '*', '/']

    def backtrack(expresion="", index=0, iter_count=0):
        iter_count += 1  # Incrementar el contador de iteraciones

        # Prueba con cada número y operador no utilizado
        for num in numeros:
            if str(num) not in expresion:

                nueva_expresion = expresion + str(num)
                resultado_parcial = eval(nueva_expresion)

                # Si ya tenemos 4 números, quiere decir que tenemos todos los operadores,
                # por lo que el resultado parcial realmente es el resultado final
                if index == 4:
                    if resultado_parcial == objetivo:
                        return nueva_expresion, iter_count
                    else:
                        continue
                
                # Si el resultado parcial no es un entero no seguimos con esta rama, ya que
                # va a ser muy poco probable que el resultado final sea un entero
                elif(resultado_parcial % 1 != 0):
                    continue
                # Si no es ninguno de los casos anteriores seguimos con el siguiente operador
                else:
                    for operador in operadores:
                        if operador not in nueva_expresion:
                            result, iter_count = backtrack(nueva_expresion + operador, index + 1, iter_count)
                            if result:
                                return result, iter_count

        return "", iter_count

    return backtrack()

# Al igual que con el algoritmo de fuerza bruta, ponemos un ejemplo en el que 
# sí se puede obtener el número objetivo y otro en el que no
numero_objetivo = 12
expresion, iteraciones = obtener_expresion_backtrack(numero_objetivo)
imprimir_resultado(expresion, iteraciones, numero_objetivo)

numero_objetivo = 200
expresion, iteraciones = obtener_expresion_backtrack(numero_objetivo)
imprimir_resultado(expresion, iteraciones, numero_objetivo)


Realizadas 106 iteraciones para obtener el número 12
Expresión para obtener el número 12: 1+2*7-9/3

Realizadas 38083 iteraciones para obtener el número 200
Dadas las restricciones del problema, no existe una expresión para obtener el número 200



In [43]:
# Comprobamos que con este nuevo algoritmo también podemos
# obtener todas las soluciones dentro del rango de números posibles
soluciones = []
for numero in range(-69,78):
    solucion, iteraciones = obtener_expresion_backtrack(numero)
    if solucion != "":
        soluciones.append(numero)
    else: 
        print(f"No se ha encontrado solución para el número {numero}")

print(f"Se han encontrado todas las soluciones posibles")


Se han encontrado todas las soluciones posibles


In [44]:
# Vamos a crear 5 números aleatorios calculables con las restricciones del problema para comparar
# el rendimiento de ambas implementaciones
import random

numeros_aleatorios = [random.randint(-69,77) for _ in range(5)]

for numero in numeros_aleatorios:
    print(f"Número objetivo: {numero}")
    expresion, iteraciones = obtener_expresion(numero)
    print(f"Realizadas {iteraciones} iteraciones por el algoritmo de fuerza bruta")
    expresion, iteraciones = obtener_expresion_backtrack(numero)
    print(f"Realizadas {iteraciones} iteraciones por el algoritmo de backtracking\n")

Número objetivo: -29
Realizadas 14073 iteraciones por el algoritmo de fuerza bruta
Realizadas 575 iteraciones por el algoritmo de backtracking

Número objetivo: 55
Realizadas 10523 iteraciones por el algoritmo de fuerza bruta
Realizadas 800 iteraciones por el algoritmo de backtracking

Número objetivo: -69
Realizadas 10661 iteraciones por el algoritmo de fuerza bruta
Realizadas 391 iteraciones por el algoritmo de backtracking

Número objetivo: -57
Realizadas 41515 iteraciones por el algoritmo de fuerza bruta
Realizadas 3641 iteraciones por el algoritmo de backtracking

Número objetivo: 58
Realizadas 84236 iteraciones por el algoritmo de fuerza bruta
Realizadas 8166 iteraciones por el algoritmo de backtracking



Tanto en esta última comparación, como en el ejemplo de uso del algoritmo, se puede ver la mejora de eficiencia de este, ya que necesita bastantes menos iteraciones para llegar al resultado objetivo. Además, en el peor de los casos, donde se le de como valor objetivo un número que no puede ser calculado con estas restricciones del problema, el algoritmo de backtracking realiza un total de 38083 iteraciones, casi 10 veces menos que las 362880 iteraciones que realiza el algoritmo por fuerza bruta.

(*)Calcula la complejidad del algoritmo

RESPUESTA:

La complejidad del nuevo algoritmo se ve reducida al ir construyendo paso a paso la operación y no tener que crear y recorrer todas las permutaciones. Además, la poda de las ramas que dan números decimales como resultados parciales nos evitan evaluar ciertas operaciones poco prometedoras, elevando aún más la eficiencia del algoritmo de backtracking. Sin embargo, al tener dos for, uno que recorre los números y otro que recorre los operadores, podría parecer que la complejidad es $$O(n \cdot m)$$ pero esto no es así, ya que dentro del for que recorre los operadores se hace una llamada recursiva a si misma, por lo que es complicado determinar cúal es la complejidad exacta del problema.

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

In [51]:
# Creamos una lista de 5 números enteros aleatorios, pero acotamos el rango para que haya
# posibilidad de obtener números enteros que se puedan obtener con las restricciones del problema

numeros_aleatorios = [random.randint(-100, 100) for _ in range(5)]

Aplica el algoritmo al juego de datos generado

In [52]:
for numero in numeros_aleatorios:
    expresion, iteraciones = obtener_expresion_backtrack(numero)
    imprimir_resultado(expresion, iteraciones, numero)

Realizadas 956 iteraciones para obtener el número -49
Expresión para obtener el número -49: 1+8/2-6*9

Realizadas 38083 iteraciones para obtener el número -91
Dadas las restricciones del problema, no existe una expresión para obtener el número -91

Realizadas 104 iteraciones para obtener el número 13
Expresión para obtener el número 13: 1+2*7-6/3

Realizadas 389 iteraciones para obtener el número -51
Expresión para obtener el número -51: 1+4/2-6*9

Realizadas 667 iteraciones para obtener el número -68
Expresión para obtener el número -68: 1+6/2-8*9



Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

RESPUESTA:

Mayoritariamente he utilizado los apuntes obtenidos en la asignatura de Matemáticas para la Inteligencia Artificial para el temario de combinatoria, para poder realizar los cálculos a priori de la cantidad de combinaciones y corroborar luego que los datos obtenidos por el código de Python eran correctos.

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

RESPUESTA:

Una posible forma de avanzar en el estudio del problema es ahondar más en posibles reglas para realizar la poda para poder utilizar alternativas a la que he utilizado yo, o incluso utilizar varias a la vez. En base a este tipo de reglas también se me ocurre implementar alguna heurística más enfocada decidir que nodos visitar y explorar primero las soluciones más prometedoras. Todas estas posibles soluciones comentadas conllevan un estudio minucioso y profundo de las características y restricciones del problema, para poder crear reglas que ayuden a mejorar el algoritmo y no interfieran negativamente en este.